lib/src/executable.dart

```diff
@@ -101,21 +112,103 @@ Future<void> handleFastFormat(List<String> args) async {
+  final runScriptContents = buildDartDevRunScriptContents();
+  if (!_runScript.existsSync() ||
+      _runScript.readAsStringSync() != runScriptContents) {
     logTimedSync(log, 'Generating run script', () {
       createCacheDir();
       _runScript.writeAsStringSync(buildDartDevRunScriptContents());
     }, level: Level.INFO);
   }
-}
 
-bool get shouldWriteRunScript =>
-    !_runScript.existsSync() ||
-    _runScript.readAsStringSync() != buildDartDevRunScriptContents();
+  // Generate a digest of inputs to the run script. We use this to determine
+  // whether we need to recompile the executable.
+  String encodedDigest;
+  logTimedSync(log, 'Computing run script digest', () {
+    var configHasRelativeImports = false;
+    var configHasSamePackageImports = false;
+    if (_config.existsSync()) {
+      final contents = _config.readAsStringSync();
+      configHasRelativeImports =
+          RegExp(r'''^import ['"][^:]+''').hasMatch(contents);
+      final currentPackageName =
+          Pubspec.parse(File('pubspec.yaml').readAsStringSync()).name;
+      configHasSamePackageImports =
+          RegExp('''import ['"]package:$currentPackageName''')
+              .hasMatch(contents);
+    }
+
+    if (configHasSamePackageImports) {
+      log.fine(
+          'Skipping compilation because $_configPath imports from its own package.');
+      // If the config imports from its own source files, we don't have a way of
+      // efficiently tracking changes that would require recompilation of this
+      // executable, so we skip the compilation altogether.
+      if (_runExecutable.existsSync()) {
+        _runExecutable.deleteSync();
+      }
+      if (_runExecutableDigest.existsSync()) {
+        _runExecutableDigest.deleteSync();
+      }
+      return;
+    }
+
+    final digest = md5.convert([
+      ..._packageConfig.readAsBytesSync(),
+      if (_config.existsSync()) ..._config.readAsBytesSync(),
+      if (configHasRelativeImports)
+        for (final file in Glob('tool/**.dart', recursive: true)
+            .listSync()
+            .whereType<File>()
+            .where((f) => f.path != _configPath))
+          ...file.readAsBytesSync(),
+    ]);
+    encodedDigest = base64.encode(digest.bytes);
+  }, level: Level.FINE);
+
+  if (encodedDigest != null &&
+      (!_runExecutableDigest.existsSync() ||
+          _runExecutableDigest.readAsStringSync() != encodedDigest)) {
+    // Digest either didn't exist or is outdated, so we (re-)compile.
+    logTimedSync(log, 'Compiling run script', () {
+      // Delete the previous exectuable and digest so that if we hit a failure
+      // trying to compile, we don't leave the outdated one in place.
+      if (_runExecutable.existsSync()) {
+        _runExecutable.deleteSync();
+      }
+      if (_runExecutableDigest.existsSync()) {
+        _runExecutableDigest.deleteSync();
+      }
+
+      final args = ['compile', 'exe', _runScriptPath, '-o', _runExecutablePath];
```

Comment:
It looks like Pub likely uses the `kernel` compile command; https://github.com/dart-lang/pub/blob/400f21e9883ce6555b66d3ef82f0b732ba9b9fc8/lib/src/dart.dart#L149-L150

I wonder why that's the one they chose...

Is there a reason you went with `exe` over any of other options? If so, could you add a comment?

Reply:
There wasn't a particular reason I chose `exe`, no, and now that I think about it, `kernel` is probably the better choice because we don't need this to be portable to systems that don't have the Dart SDK installed (we can run `dart .dart_tool/dart_dev/run.dill`). Maybe that will save some time on the compilation since (if I understand it correctly) it wouldn't need to include the SDK in the output?

Reply:
Ok so I did some benchmarking:

### `dart compile`
How long it takes to compile the generated `run.dart` script. This is the "hit" users would take on first usage for the sake of faster subsequent, cached runs.

| SDK                | Command               | Time   |
| ------------------ | --------------------- | ------ |
| 2.13.4             | `dart compile exe`    | ~10.5s |
| 2.13.4             | `dart compile kernel` | ~9.8s  |
| 2.14.4             | `dart compile exe`    | ~11.5s |
| 2.14.4             | `dart compile kernel` | ~10.5s |
| 2.15.0-268.18.beta | `dart compile exe`    | ~11.5s |
| 2.15.0-268.18.beta | `dart compile kernel` | ~10.7s |

```ad-note
Note: dart_dev is not yet null-safe. Once migrated, we assume compilation time will improve, but probably not significantly.
```

### Running compiled Dart program
How long it takes to run the output of `dart compile`. In particular, we want to know if it's faster to run an `exe` or a `dill`. These times plus the compilation times will help determine which option is best.

| SDK                | Compiled Type                    | Time   |
| ------------------ | -------------------------------- | ------ |
| 2.13.4             | exe: `./dart_dev`                | ~0.02s |
| 2.13.4             | kernel: `dart dart_dev.dill`     | ~0.5s  |
| 2.13.4             | kernel: `dart run dart_dev.dill` | ~0.8s  |
| 2.14.4             | exe: `./dart_dev`                | ~0.02s |
| 2.14.4             | kernel: `dart dart_dev.dill`     | ~0.5s  |
| 2.14.4             | kernel: `dart run dart_dev.dill` | ~0.8s  |
| 2.15.0-268.18.beta | exe: `./dart_dev`                | ~0.02s |
| 2.15.0-268.18.beta | kernel: `dart dart_dev.dill`     | ~0.5s  |
| 2.15.0-268.18.beta | kernel: `dart run dart_dev.dill` | ~0.8s  |

```ad-note
Note: there is a perf hit (~0.9s) when running a Dart exe for the first time.
```

---

The tl;dr: here is that compiling a kernel dill is about 0.8-1s faster than a native executable, but running that dill takes about 0.5s every time compared to 0.02s for native executables (excluding first run). Based on that, I'm inclined to stick with `dart compile exe` because the first run is already going to be slow; might as well optimize for the best case to get the most speedup possible on every subsequent run.